[caffe2] Allow the elimination of implicit calls to strlen when using the RECORD_FUNCTION macros

Orvid opened the pull request:
Summary:
With the way these were written, any string literals that were being passed in, like `__func__`, were only ever passed down as a `const char*`, so this switches it over to take a `std::string_view` at the deepest part.

This also has the side effect of allowing `std::string_view` to be passed to the `RECORD_FUNCTION` macros as well.

Test Plan:
contbuilds

Rollback Plan:

Differential Revision: D74681042

swolchok reviewed (changes requested):
> Test plan: contbuilds

But how do you know anything is happening at compile-time here? Since the relevant overload of extractName is in a .cpp file, it won't get inlined unless people use LTO, which is not a given in OSS PyTorch (I don't see it in our build files after a quick grep), and I don't see anything about verifying compile-time length extraction in your test plan.

Orvid commented:
> > Test plan: contbuilds
> 
> But how do you know anything is happening at compile-time here? Since the relevant overload of extractName is in a .cpp file, it won't get inlined unless people use LTO, which is not a given in OSS PyTorch (I don't see it in our build files after a quick grep), and I don't see anything about verifying compile-time length extraction in your test plan.

The existence of the overload is enough to do what is needed. It doesn't need LTO, and the only inlining it needs is the trivial constructor for `std::string_view`. The existence of the overload causes the conversion to `std::string_view` to occur in the caller, rather than the callee's callee's callee, and the caller will already have the string literal present if it's going to exist, meaning the `strlen` call from the inlined constructor to `std::string_view` will get evaluated at compile time at pretty much any optimization level.

swolchok reviewed (approved):
OK I buy that argument, objection withdrawn

Orvid commented:
After looking deeper at this, it looks like I'm going to have to drop the approach of guaranteeing that any callers get converted, and will instead just have to settle for just making it accept a `std::string_view` at the innermost level, since doing anything else would require duplicating far to many functions :(

Orvid committed:
[caffe2] Allow the elimination of implicit calls to strlen when using the RECORD_FUNCTION macros (#153567)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153567

With the way these were written, any string literals that were being passed in, like `__func__`, were only ever passed down as a `const char*`, so this switches it over to take a `std::string_view` at the deepest part.

This also has the side effect of allowing `std::string_view` to be passed to the `RECORD_FUNCTION` macros as well.

Test Plan:
contbuilds

Rollback Plan:

Reviewed By: swolchok

Differential Revision: D74681042

Orvid commented:
The only checks I'm seeing that are marked as failed are ones that had infra failures. Given that the only errors on the prior update to the diff were lint failures due to me not having run clang-format, I went ahead and landed it, but it seems like the merge bot wasn't happy about that :(

huydhn commented:
@pytorchbot merge -f 'The diff has been landed internally, lots of infra failures here'
